### PR TITLE
Continuous Documentation

### LINKED ISSUE #342: Automated documentation previews in pull requests
**Description of the desired feature**

Make it easier to preview changes to the documentation **online** in every pull request, rather than having to clone the branch and do a **local** build. I.e. Continuous Documentation.

Will be especially helpful for new contributors and even reviewers wanting a quick look at what's going up on the website. This was hinted at during #314.

Documentation preview

I'm gonna try to set up an automated Continuous Deployment site, just a heads up!

**Are you willing to help implement and maintain this feature?** Yes

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

Making it easier to preview documentation changes in Pull Requests via 'Continuous Documentation'. Currently deployed to serverless infrastructure on [Zeit Now](https://zeit.co/now).

Zeit Now is meant to be a zero-config/serverless, but that's for the javascript world. I've hacked the `package.json` to get `sphinx` and `pygmt` up and running to build the docs. The static files are then copied to a `public` folder where it gets deployed automatically to a url that looks like `pygmt-abcdef.now.sh`. You can see an example at https://pygmt-l8nmyn8mx.now.sh/.

Merge after #343, Fixes #342

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

### CODE REVIEW

**Comment by weiji14** [comment]
Ok, this is almost ready! Previews of the website can now viewed by clicking on the 'View deployment' button:
image

The bot is a lot noisier than I expected, but I've toggled on a [silent](https://zeit.co/docs/configuration/#configuration/github-silent) flag that should :crossed_fingers: stop it from commenting on Pull Requests and commits once this PR is merged in.

Another thing worth considering is the related [Deploy Summary Integration](https://zeit.co/blog/deploy-summary) app (see [review](https://medium.com/front10/lets-try-build-summary-integration-by-zeit-now-a4529c4d7685)) that would detect changes to the pages and post a screenshot and direct link to the said changes. This will need to be installed (and might add more noise to our PRs) so looking to see what everyone thinks.

**Comment by leouieda** [comment]
@weiji14 thanks for taking the time! This looks really cool. Is the now service building the docs or are they being deployed from the CI? Since all our CIs build the docs anyway, would it be easier to deploy from there? The problem is that keep these build systems alive takes quite a bit of time and effort, so it would be good to explore all options before adding a new one.

**Comment by weiji14** [comment]
> @weiji14 thanks for taking the time! This looks really cool. Is the now service building the docs or are they being deployed from the CI?

Yes, the now service is building them using `npm`.

> Since all our CIs build the docs anyway, would it be easier to deploy from there? The problem is that keep these build systems alive takes quite a bit of time and effort, so it would be good to explore all options before adding a new one.

I hinted at https://github.com/GenericMappingTools/pygmt/issues/342#issuecomment-541461105 that this is indeed possible, but would require setting a token/secret variable. We could probably disable those documentation builds in a separate PR to speed them up?

The thing is, each of the built documentation sites have to be hosted online somewhere and given a unique url, which none of our current CIs are able to provide. Github Pages does do a build but is limited to one url. Another option I've looked at is [netlify](https://www.netlify.com/docs/continuous-deployment/) that does CD as well, but their pricing structure isn't as good as Now's.

**Comment by xibalba01** [comment]
Have you guys ever thought of migrating to gitlab? They have the possibility of ci/cd stuff directly built in so that you don't have to use different external apps that one has to somehow connect to github... Have a look here: https://docs.gitlab.com/ee/ci/introduction/index.html

**Comment by leouieda** [comment]
>  We could probably disable those documentation builds in a separate PR to speed them up?

I’m not too keen on that since he docs builds are a sort of test. But to all builds need to do this (probably only one per OS would be good).

> The thing is, each of the built documentation sites have to be hosted online somewhere and given a unique url, which none of our current CIs are able to provide. Github Pages does do a build but is limited to one url. Another option I've looked at is netlify that does CD as well, but their pricing structure isn't as good as Now's.

And using a key won’t work for forks because they aren’t decrypted for these builds (rightly so). But there are ways of storing build artefacts in Azure and Travis (which could be the html).

> have you guys ever thought of migrating to gitlab

Yes but the problem is that everyone is here on GitHub, which makes it much easier to recruit contributors. But it is tempting.

**Comment by weiji14** [comment]
Agreed that the documentation builds (which creates figures) does provide provide some 'integration' testing. Also with the Gitlab argument, Github does have [Github Actions](https://github.com/features/actions) now which can be used to do CI/CD, but again, this topic is getting sidetracked too much...

Are there any outstanding issues with this PR @leouieda, or is it ready for approval/merge? I'm actually waiting to get this merged in to start working on some gallery examples. The workshop is coming up soon...

**Comment by leouieda** [comment]
@weiji14 would you mind inserting some comments in the two files included here about what they are? Also, did this setup require signing up for these services? If so, were they done with your personal account? Would you mind posting some more information about this here so we can I have a record in case someone else need to manage this in the future? One thing we want to avoid is having a single person be able to access/manage things (which reminds me, we need to add you to Azure and PyPI).

**Comment by leouieda** [comment]
Sorry for holding this up but PRs are meant for discussion before we commit to a given feature. It’s much easier to add things than it is to remove them later (granted, not in the case of this PR). I want to encourage a culture of deeper discussion of things before giving the 👍 and merging, even if that means going a bit slower in PyGMT than GMT core.

**Comment by weiji14** [comment]
No don't apologize! I totally understand from a maintenance point of view (having worked on >10 year old legacy code before). A quick search shows that it isn't easy to add comments json files (see https://stackoverflow.com/questions/244777/can-comments-be-used-in-json) but I'll look around a bit more.

In terms of the Now account setup, I've set up a 'gmt' organization and I'm currently the 'owner', with @seisman being a user. I'll send an invite to your email too actually.

**Comment by leouieda** [comment]
@weiji14 i thought that might be the case. Don’t worry too much about it. What you can do is include a section in `MAINTENANCE.md` which briefly explains the setup, files involved, and links to services so we can easily find them.

File: MAINTENANCE.md [hunk 1]
```diff
@@ -61,6 +61,15 @@ repository to avoid duplicating work across multiple repositories.
...
 
+## Continuous Documentation
+
+We use the [Zeit Now for Github integration](https://zeit.co/github) to preview changes
+made to our documentation website every time we make a commit in a pull request.
+The integration service has a configuration file `now.json`, with a list of options to
+change the default behaviour at https://zeit.co/docs/configuration.
+The actual script `package.json` is used by Zeit Now to install the necessary packages,
+build the documentation, copy the files to a 'public' folder and deploy that to the web,
+see https://zeit.co/docs/v2/build-step/?query=package.json#defining-a-build-script.
```
Comment (weiji14) [inline MAINTENANCE.md]: > @weiji14 i thought that might be the case. Don’t worry too much about it. What you can do is include a section in `MAINTENANCE.md` which briefly explains the setup, files involved, and links to services so we can easily find them.

Ok, the comments in JSON definitely doesn't work, so I've added this 'Continuous Documentation' section to 'MAINTENANCE.md', just below 'Continuous Integration'. Hopefully that's enough for now.

**Pull request merged**